File: app/src/main/java/com/mainstreetcode/teammate/repository/ModelRepo.java
package com.mainstreetcode.teammate.repository;


import android.annotation.SuppressLint;
import android.webkit.MimeTypeMap;

import com.mainstreetcode.teammate.model.Message;
import com.mainstreetcode.teammate.model.Model;
import com.mainstreetcode.teammate.persistence.EntityDao;
import com.mainstreetcode.teammate.util.ErrorHandler;

import java.io.File;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import androidx.annotation.Nullable;
import io.reactivex.Completable;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.Single;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;
import okhttp3.MediaType;
import okhttp3.MultipartBody;
import okhttp3.RequestBody;
import retrofit2.HttpException;

import static io.reactivex.Maybe.concatDelayError;

/**
 * Repository that manages {@link Model} CRUD operations
 */

public abstract class ModelRepo<T extends Model<T>> {

    static final int DEF_QUERY_LIMIT = 12;

    private final Function<List<T>, List<T>> saveListFunction = provideSaveManyFunction();
    private final Function<T, T> saveFunction = model -> saveListFunction.apply(Collections.singletonList(model)).get(0);

    public abstract EntityDao<? super T> dao();

    public abstract Single<T> createOrUpdate(T model);

    public abstract Flowable<T> get(String id);

    public abstract Single<T> delete(T model);

    abstract Function<List<T>, List<T>> provideSaveManyFunction();

    public final Flowable<T> get(T model) {
        return model.isEmpty()
                ? Flowable.error(new IllegalArgumentException("Model does not exist"))
                : get(model.getId()).map(getLocalUpdateFunction(model));
    }

    @SuppressLint("CheckResult")
    @SuppressWarnings("ResultOfMethodCallIgnored")
    public final void queueForLocalDeletion(T model) {
        Completable.fromRunnable(() -> deleteLocally(model))
                .subscribeOn(Schedulers.io())
                .subscribe(() -> {}, ErrorHandler.EMPTY);
    }

    final Function<T, T> getLocalUpdateFunction(T original) {
        return emitted -> {
            original.update(emitted);
            return original;
        };
    }

    final Function<List<T>, List<T>> getSaveManyFunction() {
        return saveListFunction;
    }

    final Function<T, T> getSaveFunction() {
        return saveFunction;
    }

    final Function<List<T>, List<T>> saveAsNested() {
        return models -> {
            if (models.isEmpty()) return models;

            T litmus = models.get(0);
            if (litmus.hasMajorFields()) return saveListFunction.apply(models);

            dao().insert(Collections.unmodifiableList(models));
            return models;
        };
    }

    final Flowable<T> fetchThenGetModel(Maybe<T> local, Maybe<T> remote) {
        AtomicReference<T> reference = new AtomicReference<>();
        local = local.doOnSuccess(reference::set);
        remote = remote.doOnError(throwable -> deleteInvalidModel(reference.get(), throwable));

        return fetchThenGet(local, remote);
    }

    final void deleteInvalidModel(T model, Throwable throwable) {
        if (model == null || !(throwable instanceof HttpException)) return;

        Message message = new Message((HttpException) throwable);
        if (message.isInvalidObject() || message.isIllegalTeamMember()) deleteLocally(model);
    }

    T deleteLocally(T model) {
        dao().delete(model);
        return model;
    }

    @Nullable
    MultipartBody.Part getBody(CharSequence path, String photoKey) {
        String pathString = path.toString();
        File file = new File(pathString);

        if (!file.exists()) return null;
        String extension = MimeTypeMap.getFileExtensionFromUrl(pathString);

        if (extension == null) return null;
        String type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);

        if (type == null) return null;
        RequestBody requestBody = RequestBody.create(MediaType.parse(type), file);

        return MultipartBody.Part.createFormData(photoKey, file.getName(), requestBody);
    }

    static <R> Flowable<R> fetchThenGet(Maybe<R> local, Maybe<R> remote) {
        return concatDelayError(Arrays.asList(local, remote));
    }

}

File: app/src/main/java/com/mainstreetcode/teammate/repository/ModelRepo.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.repository


import android.annotation.SuppressLint
import android.webkit.MimeTypeMap
import com.mainstreetcode.teammate.model.Message
import com.mainstreetcode.teammate.model.Model
import com.mainstreetcode.teammate.persistence.EntityDao
import com.mainstreetcode.teammate.util.ErrorHandler
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Maybe
import io.reactivex.Maybe.concatDelayError
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers
import okhttp3.MediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody
import retrofit2.HttpException
import java.io.File
import java.util.concurrent.atomic.AtomicReference

/**
 * Repository that manages [Model] CRUD operations
 */

abstract class ModelRepo<T : Model<T>> {

    @Suppress("LeakingThis")
    internal val saveManyFunction = provideSaveManyFunction()
    internal val saveFunction = { model: T -> saveManyFunction.invoke(listOf(model))[0] }

    abstract fun dao(): EntityDao<in T>

    abstract fun createOrUpdate(model: T): Single<T>

    abstract operator fun get(id: String): Flowable<T>

    abstract fun delete(model: T): Single<T>

    internal abstract fun provideSaveManyFunction(): (List<T>) -> List<T>

    operator fun get(model: T): Flowable<T> = when {
        model.isEmpty -> Flowable.error(IllegalArgumentException("Model does not exist"))
        else -> get(model.id).map(getLocalUpdateFunction(model))
    }

    @SuppressLint("CheckResult")
    fun queueForLocalDeletion(model: T) {
        Completable.fromRunnable { deleteLocally(model) }
                .subscribeOn(Schedulers.io())
                .subscribe({ }, ErrorHandler.EMPTY::invoke)
    }

    internal fun getLocalUpdateFunction(original: T): (T) -> T = { original.update(it); original }

    internal fun saveAsNested(): (List<T>) -> List<T> = inner@{ models ->
        if (models.isEmpty()) return@inner models

        val parted = models.partition { it.isEmpty || it.hasMajorFields() }

        saveManyFunction.invoke(parted.first) // Nested save empty models or complete ones with upsert
        dao().insert(parted.second) // For non empty models with incomplete data, insert, if a copy exists, it will be rejected

        models
    }


    internal fun fetchThenGetModel(localSource: Maybe<T>, remoteSource: Maybe<T>): Flowable<T> {
        var local = localSource
        var remote = remoteSource
        val reference = AtomicReference<T>()
        local = local.doOnSuccess(reference::set)
        remote = remote.doOnError { throwable -> deleteInvalidModel(reference.get(), throwable) }

        return fetchThenGet(local, remote)
    }

    internal fun deleteInvalidModel(model: T?, throwable: Throwable) {
        if (model == null || throwable !is HttpException) return

        val message = Message(throwable)
        if (message.isInvalidObject || message.isIllegalTeamMember) deleteLocally(model)
    }

    internal open fun deleteLocally(model: T): T {
        dao().delete(model)
        return model
    }

    internal fun getBody(path: CharSequence?, photoKey: String): MultipartBody.Part? {
        path ?: return null

        val pathString = path.toString()
        val file = File(pathString)

        if (!file.exists()) return null
        val extension = MimeTypeMap.getFileExtensionFromUrl(pathString) ?: return null

        val type = MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension) ?: return null

        val requestBody = RequestBody.create(MediaType.parse(type), file)

        return MultipartBody.Part.createFormData(photoKey, file.name, requestBody)
    }

    companion object {

        internal const val DEF_QUERY_LIMIT = 12

        internal fun <R> fetchThenGet(local: Maybe<R>, remote: Maybe<R>): Flowable<R> =
                concatDelayError(listOf(local, remote))
    }

}
